File: mobile/ios/Sources/ProfileService.swift
import Foundation

/// The signed-in user's editable profile, returned by `GET /api/local/profile`
/// and updated via `PATCH /api/local/profile`. All fields are optional so the
/// screen renders even when the backend has only some of them filled in.
struct Profile: Codable, Equatable {
    var fullName: String?
    var age: Int?
    var idDocument: String?
    var phone: String?
    var email: String?
    /// Free-text "about me" blurb shown under the name on the profile screen.
    var bio: String?
    /// Avatar source — either an `http(s)://` URL or an inline `data:image/jpeg;base64,…`
    /// data URL produced by the avatar picker. `nil` falls back to initials.
    var avatarURL: String?
    /// Identity-verification state from `verification_status`:
    /// "unverified" | "pending" | "verified" | "rejected". Defaults to
    /// "unverified" when the backend omits it — drives the "Verify your
    /// identity" card on the profile.
    var verificationStatus: String

    enum CodingKeys: String, CodingKey {
        case age, phone, email, bio
        case fullName = "full_name"
        case idDocument = "id_document"
        case avatarURL = "avatar_url"
        case verificationStatus = "verification_status"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        fullName = try c.decodeIfPresent(String.self, forKey: .fullName)
        age = try c.decodeIfPresent(Int.self, forKey: .age)
        idDocument = try c.decodeIfPresent(String.self, forKey: .idDocument)
        phone = try c.decodeIfPresent(String.self, forKey: .phone)
        email = try c.decodeIfPresent(String.self, forKey: .email)
        bio = try c.decodeIfPresent(String.self, forKey: .bio)
        avatarURL = try c.decodeIfPresent(String.self, forKey: .avatarURL)
        verificationStatus = (try c.decodeIfPresent(String.self, forKey: .verificationStatus))
            .flatMap { $0.isEmpty ? nil : $0 } ?? "unverified"
    }

    /// Memberwise initializer (kept because the custom `init(from:)` above
    /// suppresses the synthesized one). Used by `updateProfile`'s fallback echo.
    init(
        fullName: String?,
        age: Int?,
        idDocument: String?,
        phone: String?,
        email: String?,
        bio: String?,
        avatarURL: String?,
        verificationStatus: String = "unverified"
    ) {
        self.fullName = fullName
        self.age = age
        self.idDocument = idDocument
        self.phone = phone
        self.email = email
        self.bio = bio
        self.avatarURL = avatarURL
        self.verificationStatus = verificationStatus
    }
}

/// Networking for the signed-in user's profile against the local Next.js API.
/// Mirrors `BookingService`/`HostService`: pure URLSession + Codable, reading the
/// bearer token straight from `UserDefaults` under `AuthStore.tokenKey`.
///
///   GET   {base}/api/local/profile  (Bearer qk_token) → Profile
///   PATCH {base}/api/local/profile  (Bearer qk_token) { full_name, age, id_document, phone, bio, avatar_url } → Profile
struct ProfileService {
    static let shared = ProfileService()

    private let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 15
        cfg.waitsForConnectivity = true
        return URLSession(configuration: cfg)
    }()

    /// The persisted bearer token, or `nil` when browsing as a guest.
    var token: String? {
        let value = UserDefaults.standard.string(forKey: AuthStore.tokenKey)
        return (value?.isEmpty == false) ? value : nil
    }

    // MARK: - Read

    /// Load the signed-in user's profile.
    func fetchProfile() async throws -> Profile {
        guard let token else { throw ProfileError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/profile")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ProfileError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ProfileError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            throw ProfileError.message(Self.decodeError(data) ?? "Couldn't load your profile (\(http.statusCode)).")
        }
        return try JSONDecoder().decode(Profile.self, from: data)
    }

    // MARK: - Update

    /// Save the editable fields. Sends `{ full_name, age, phone, bio, avatar_url }`;
    /// `age`/`bio`/`avatar_url` are sent as JSON null when cleared. Returns the
    /// updated profile when the backend echoes one, otherwise the values just sent.
    ///
    /// `id_document` is deliberately NOT sent. It is identity, and the endpoint now
    /// refuses any value that differs from what is stored — changing it goes through
    /// `requestIDChange` and an operator's approval. Sending the unchanged value would
    /// be accepted, but there is no reason to put a person's ID number on the wire on
    /// every bio edit.
    @discardableResult
    func updateProfile(
        fullName: String,
        age: Int?,
        phone: String,
        bio: String,
        avatarURL: String?
    ) async throws -> Profile {
        guard let token else { throw ProfileError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/profile")!
        var request = URLRequest(url: url)
        request.httpMethod = "PATCH"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        var body: [String: Any] = [
            "full_name": fullName,
            "phone": phone,
        ]
        // Send the age as a number when set, explicit null when cleared.
        body["age"] = age ?? NSNull()
        // Send the bio as text, or explicit null once emptied.
        body["bio"] = bio.isEmpty ? NSNull() : bio
        // Send the avatar (http URL or data: URL), or explicit null when removed.
        body["avatar_url"] = avatarURL ?? NSNull()
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ProfileError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ProfileError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            throw ProfileError.message(Self.decodeError(data) ?? "Couldn't save your profile (\(http.statusCode)).")
        }
        // Prefer the server's echo; fall back to the submitted values.
        if let updated = try? JSONDecoder().decode(Profile.self, from: data) {
            return updated
        }
        // The echo is what carries id_document back; without one there is nothing
        // truthful to put here, so it stays nil rather than inventing a value.
        return Profile(
            fullName: fullName,
            age: age,
            idDocument: nil,
            phone: phone,
            email: nil,
            bio: bio.isEmpty ? nil : bio,
            avatarURL: avatarURL,
            verificationStatus: "unverified"
        )
    }

    // MARK: - Change password

    /// Change the signed-in user's password
    /// (`POST /api/local/change-password` (Bearer) `{ current_password,
    /// new_password }` → `{ ok: true }`). Throws `ProfileError.message` carrying
    /// the server `{ error }` on a 400 (e.g. wrong current password).
    func changePassword(currentPassword: String, newPassword: String) async throws {
        guard let token else { throw ProfileError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/change-password")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "current_password": currentPassword,
            "new_password": newPassword,
        ])

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ProfileError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ProfileError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            throw ProfileError.message(Self.decodeError(data) ?? "Couldn't change your password (\(http.statusCode)).")
        }
    }

    // MARK: - Helpers

    private static func decodeError(_ data: Data) -> String? {
        struct ErrorBody: Decodable { let error: String }
        return (try? JSONDecoder().decode(ErrorBody.self, from: data))?.error
    }

    // MARK: - ID change requests

    /// The current ID number and the state of any request to change it
    /// (`GET /api/local/profile/id-change` (Bearer)).
    func fetchIDChangeState() async throws -> IDChangeState {
        guard let token else { throw ProfileError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/profile/id-change")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ProfileError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ProfileError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            throw ProfileError.message(Self.decodeError(data) ?? "Couldn't load your ID request (\(http.statusCode)).")
        }
        return try JSONDecoder().decode(IDChangeState.self, from: data)
    }

    /// Ask for the ID number on the profile to be changed
    /// (`POST /api/local/profile/id-change` (Bearer)
    /// `{ requested_value, doc_type, front, back?, reason? }`).
    ///
    /// The front image is required by the server: without a document the reviewer has
    /// nothing to check the typed number against. Resubmitting replaces a request that
    /// is still waiting rather than queueing a second one.
    @discardableResult
    func requestIDChange(
        requestedValue: String,
        docType: String,
        front: String,
        back: String?,
        reason: String
    ) async throws -> IDChangeState {
        guard let token else { throw ProfileError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/profile/id-change")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        var body: [String: Any] = [
            "requested_value": requestedValue,
            "doc_type": docType,
            "front": front,
        ]
        if let back, !back.isEmpty { body["back"] = back }
        let trimmedReason = reason.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedReason.isEmpty { body["reason"] = trimmedReason }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ProfileError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ProfileError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            // A 400 here carries the core's own wording ("A national ID number is 14
            // digits"), which is exactly what the user needs to see.
            throw ProfileError.message(Self.decodeError(data) ?? "Couldn't send your request (\(http.statusCode)).")
        }
        return try JSONDecoder().decode(IDChangeState.self, from: data)
    }

    /// Withdraw a request that is still awaiting review
    /// (`DELETE /api/local/profile/id-change` (Bearer)).
    @discardableResult
    func cancelIDChangeRequest() async throws -> IDChangeState {
        guard let token else { throw ProfileError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/profile/id-change")!
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ProfileError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ProfileError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            throw ProfileError.message(Self.decodeError(data) ?? "Couldn't withdraw your request (\(http.statusCode)).")
        }
        return try JSONDecoder().decode(IDChangeState.self, from: data)
    }
}

// MARK: - ID change models

/// Where a request to change the profile's ID number has got to.
enum IDChangeStatus: String, Codable {
    case pending, approved, rejected
}

/// One request to change the ID number, as the server reports it back.
struct IDChangeRequest: Codable, Equatable {
    var id: String
    var status: IDChangeStatus
    var requestedValue: String
    var currentValue: String?
    var docType: String
    var reason: String?
    /// The operator's note. On a rejection this is the reason to show the user.
    var notes: String?
    var submittedAt: String?
    var reviewedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, status, reason, notes
        case requestedValue = "requested_value"
        case currentValue = "current_value"
        case docType = "doc_type"
        case submittedAt = "submitted_at"
        case reviewedAt = "reviewed_at"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(String.self, forKey: .id) ?? ""
        // An unknown status is treated as pending: it is the state that shows the
        // safest thing (waiting, action hidden) rather than offering a resubmit that
        // the server would refuse. Decoded via the raw string so a value this build
        // does not know about degrades instead of failing the whole response.
        let rawStatus = try c.decodeIfPresent(String.self, forKey: .status)
        status = rawStatus.flatMap(IDChangeStatus.init(rawValue:)) ?? .pending
        requestedValue = try c.decodeIfPresent(String.self, forKey: .requestedValue) ?? ""
        currentValue = try c.decodeIfPresent(String.self, forKey: .currentValue)
        docType = try c.decodeIfPresent(String.self, forKey: .docType) ?? "national_id"
        reason = try c.decodeIfPresent(String.self, forKey: .reason)
        notes = try c.decodeIfPresent(String.self, forKey: .notes)
        submittedAt = try c.decodeIfPresent(String.self, forKey: .submittedAt)
        reviewedAt = try c.decodeIfPresent(String.self, forKey: .reviewedAt)
    }
}

/// The ID number on the profile plus whatever became of the latest request for it.
struct IDChangeState: Codable, Equatable {
    /// The value on the profile right now — the only one that counts.
    var current: String?
    var request: IDChangeRequest?
    /// False only while a request is waiting; the screen hides the action on false.
    var canRequest: Bool

    enum CodingKeys: String, CodingKey {
        case current, request
        case canRequest = "can_request"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        current = try c.decodeIfPresent(String.self, forKey: .current)
        request = try c.decodeIfPresent(IDChangeRequest.self, forKey: .request)
        canRequest = try c.decodeIfPresent(Bool.self, forKey: .canRequest) ?? true
    }

    init(current: String?, request: IDChangeRequest?, canRequest: Bool) {
        self.current = current
        self.request = request
        self.canRequest = canRequest
    }
}

/// The identity documents a change request may be filed against. Mirrors DOC_TYPES
/// in the backend's host-verification-core.ts, so a request and a verification always
/// mean the same thing by 'passport'.
enum IDDocumentType: String, CaseIterable, Identifiable {
    case nationalID = "national_id"
    case passport
    case residencePermit = "residence_permit"

    var id: String { rawValue }

    var labelKey: String {
        switch self {
        case .nationalID:      return "idChange.docType.nationalId"
        case .passport:        return "idChange.docType.passport"
        case .residencePermit: return "idChange.docType.residencePermit"
        }
    }
}

/// Errors surfaced to the profile-settings UI.
enum ProfileError: LocalizedError {
    case notSignedIn
    case message(String)

    var errorDescription: String? {
        switch self {
        case .notSignedIn:       return "Sign in to edit your profile"
        case let .message(text): return text
        }
    }
}

File: mobile/android/app/src/main/java/com/quickin/app/ProfileService.kt
package com.quickin.app

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * The editable profile fields shown on the profile-settings screen
 * (`GET /api/local/profile`). All optional — any may be blank/absent server-side.
 */
data class Profile(
    val fullName: String = "",
    val email: String = "",
    val age: Int? = null,
    val idDocument: String = "",
    val phone: String = "",
    /** Free-text "about me" blurb. Blank when unset server-side. */
    val bio: String = "",
    /**
     * Avatar image source: an `http(s)` URL or an inline `data:image/...;base64,…` data URL
     * (Coil's [coil.compose.AsyncImage] decodes both). Null when the user has no photo.
     */
    val avatarUrl: String? = null,
    /**
     * Identity-verification state for this account (parsed from "verification_status";
     * defaults to "unverified"): "unverified" | "pending" | "verified" | "rejected". Drives the
     * "Verify your identity" card on the profile-settings screen.
     */
    val verificationStatus: String = "unverified"
)

/**
 * One request to change the ID number on the profile, as the server reports it back.
 * [notes] is the operator's note — on a rejection it is the reason to show the user.
 */
data class IdChangeRequest(
    val id: String = "",
    /** "pending" | "approved" | "rejected". */
    val status: String = "pending",
    val requestedValue: String = "",
    val docType: String = "national_id",
    val reason: String = "",
    val notes: String = ""
)

/**
 * The ID number on file plus whatever became of the latest request for it.
 * [canRequest] is false only while a request is waiting — the screen hides the action then.
 */
data class IdChangeState(
    val current: String = "",
    val request: IdChangeRequest? = null,
    val canRequest: Boolean = true
)

/**
 * The identity documents a change request may be filed against. Mirrors DOC_TYPES in the
 * backend's host-verification-core.ts, so a request and a verification always mean the same
 * thing by 'passport'. [labelRes] is resolved by the screen, not here.
 */
enum class IdDocumentType(val key: String, val labelRes: Int) {
    NATIONAL_ID("national_id", R.string.id_change_doc_national),
    PASSPORT("passport", R.string.id_change_doc_passport),
    RESIDENCE_PERMIT("residence_permit", R.string.id_change_doc_residence)
}

/**
 * Minimal HTTP client for the signed-in user's profile. Mirrors [BookingService] /
 * [ServiceService]: HttpURLConnection + org.json on Dispatchers.IO, bearer-token auth.
 *
 *   GET   {base}/api/local/profile  -> { full_name, email, age, id_document, phone, bio, avatar_url }
 *   PATCH {base}/api/local/profile  { full_name, age, phone, bio, avatar_url } -> updated profile
 *   GET/POST/DELETE {base}/api/local/profile/id-change -> the ID-number change request
 *
 * `id_document` is READ-ONLY: the PATCH above no longer sends it and the server refuses any
 * value that differs from what is stored. It used to be an ordinary editable field, which meant
 * any account could rewrite its own identity number with nobody reviewing it. Changing it now
 * means filing a request with a photo of the document, which an operator approves.
 */
object ProfileService {

    /** Thrown so callers can distinguish "sign in" (401) from validation (400). */
    class HttpError(val code: Int, message: String) : RuntimeException(message)

    /** Loads the signed-in user's editable profile. Throws [HttpError] (401 when signed out). */
    suspend fun fetchProfile(token: String): Profile = withContext(Dispatchers.IO) {
        val text = get(token, "/api/local/profile")
        parseProfile(JSONObject(text))
    }

    /**
     * Saves the editable profile fields (`PATCH /api/local/profile`). [age] is sent as JSON null
     * when null. [avatarUrl] is sent as JSON null when null (clears the photo) — otherwise the
     * `http(s)` URL or `data:image/jpeg;base64,…` data URL the edit screen produced. Returns the
     * updated [Profile]. Throws [HttpError] (401 / 400 on validation).
     */
    suspend fun updateProfile(
        token: String,
        fullName: String,
        age: Int?,
        phone: String,
        bio: String,
        avatarUrl: String?
    ): Profile = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("full_name", fullName.trim())
            if (age != null) put("age", age) else put("age", JSONObject.NULL)
            put("phone", phone.trim())
            put("bio", bio.trim())
            if (avatarUrl != null) put("avatar_url", avatarUrl) else put("avatar_url", JSONObject.NULL)
        }
        val text = send("PATCH", token, "/api/local/profile", body)
        parseProfile(JSONObject(text))
    }

    /**
     * Changes the signed-in user's password (`POST /api/local/change-password`). Sends the
     * [currentPassword] (verified server-side) and the [newPassword]. Returns Unit on the 200
     * `{ok:true}`; throws [HttpError] on 400 (wrong current password / weak new password) or
     * 401 (signed out), carrying the server's `{error}` message.
     */
    suspend fun changePassword(
        token: String,
        currentPassword: String,
        newPassword: String
    ): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("current_password", currentPassword)
            put("new_password", newPassword)
        }
        send("POST", token, "/api/local/change-password", body)
    }

    // ---- ID change requests ---------------------------------------------------

    /** The current ID number and the state of any request to change it. */
    suspend fun fetchIdChangeState(token: String): IdChangeState = withContext(Dispatchers.IO) {
        parseIdChangeState(JSONObject(get(token, "/api/local/profile/id-change")))
    }

    /**
     * Asks for the ID number on the profile to be changed.
     *
     * [front] is required by the server — without a photo of the document the reviewer has
     * nothing to check the typed number against. [back] and [reason] are optional. Resubmitting
     * replaces a request that is still waiting rather than queueing a second one.
     *
     * Validation of the number itself is left to the server on purpose: the rules live in one
     * shared core that both the mobile API and the admin console read, and a copy here would be
     * a third place for them to drift. A 400 carries that core's own wording.
     */
    suspend fun requestIdChange(
        token: String,
        requestedValue: String,
        docType: String,
        front: String,
        back: String?,
        reason: String
    ): IdChangeState = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("requested_value", requestedValue.trim())
            put("doc_type", docType)
            put("front", front)
            if (!back.isNullOrBlank()) put("back", back)
            if (reason.isNotBlank()) put("reason", reason.trim())
        }
        parseIdChangeState(JSONObject(send("POST", token, "/api/local/profile/id-change", body)))
    }

    /** Withdraws a request that is still awaiting review. */
    suspend fun cancelIdChange(token: String): IdChangeState = withContext(Dispatchers.IO) {
        parseIdChangeState(JSONObject(sendNoBody("DELETE", token, "/api/local/profile/id-change")))
    }

    private fun parseIdChangeState(o: JSONObject): IdChangeState {
        val requestObj = o.optJSONObject("request")
        return IdChangeState(
            current = o.optStringOrNull("current").orEmpty(),
            request = requestObj?.let {
                IdChangeRequest(
                    id = it.optString("id"),
                    // An unknown status reads as pending: that state hides the action rather
                    // than offering a resubmit the server would refuse.
                    status = it.optStringOrNull("status") ?: "pending",
                    requestedValue = it.optStringOrNull("requested_value").orEmpty(),
                    docType = it.optStringOrNull("doc_type") ?: "national_id",
                    reason = it.optStringOrNull("reason").orEmpty(),
                    notes = it.optStringOrNull("notes").orEmpty()
                )
            },
            canRequest = if (o.has("can_request")) o.optBoolean("can_request", true) else true
        )
    }

    // ---- HTTP helpers (mirror BookingService) ---------------------------------

    private fun get(token: String, path: String): String {
        val conn = (URL("${Config.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Authorization", "Bearer $token")
        }
        try {
            val code = conn.responseCode
            val text = readBody(conn, code)
            if (code !in 200..299) throw HttpError(code, extractError(text, code))
            return text
        } finally {
            conn.disconnect()
        }
    }

    private fun send(method: String, token: String, path: String, body: JSONObject): String {
        val conn = (URL("${Config.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 15_000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Authorization", "Bearer $token")
        }
        try {
            conn.outputStream.use { out -> out.write(body.toString().toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            val text = readBody(conn, code)
            if (code !in 200..299) throw HttpError(code, extractError(text, code))
            return text
        } finally {
            conn.disconnect()
        }
    }

    /** Like [send] but with no request body — the DELETE that withdraws a request. */
    private fun sendNoBody(method: String, token: String, path: String): String {
        val conn = (URL("${Config.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Authorization", "Bearer $token")
        }
        try {
            val code = conn.responseCode
            val text = readBody(conn, code)
            if (code !in 200..299) throw HttpError(code, extractError(text, code))
            return text
        } finally {
            conn.disconnect()
        }
    }

    private fun readBody(conn: HttpURLConnection, code: Int): String {
        val stream = if (code in 200..299) conn.inputStream else conn.errorStream
        return stream?.bufferedReader()?.use { it.readText() }.orEmpty()
    }

    private fun extractError(text: String, code: Int): String {
        val parsed = runCatching { JSONObject(text).optString("error") }.getOrNull()
        return if (!parsed.isNullOrBlank()) parsed else "Request failed ($code)"
    }

    /**
     * Parses the profile object. Some responses nest the row under a `profile`/`user` key, so
     * unwrap that first. Accepts a couple of alternate field names for the ID document and age.
     */
    private fun parseProfile(raw: JSONObject): Profile {
        val o = raw.optJSONObject("profile") ?: raw.optJSONObject("user") ?: raw
        // optStringOrNull, not optString: an account with no ID on file has `id_document: null`,
        // and optString hands back the literal "null", which Edit Profile then printed as the
        // value instead of falling through to "Not on file".
        val idDoc = o.optStringOrNull("id_document")
            ?: o.optStringOrNull("id_passport")
            ?: o.optStringOrNull("passport")
            ?: ""
        val ageValue = if (o.has("age") && !o.isNull("age")) o.optInt("age").takeIf { it > 0 } else null
        val avatar = if (o.has("avatar_url") && !o.isNull("avatar_url")) {
            o.optStringOrNull("avatar_url")
        } else null
        return Profile(
            fullName = o.optString("full_name").ifBlank { o.optString("name") },
            email = o.optString("email"),
            age = ageValue,
            idDocument = idDoc,
            // optString returns the literal "null" for a JSON null, so guard with isNull
            // first — otherwise an unset phone/bio renders as the text "null".
            phone = if (o.isNull("phone")) "" else o.optString("phone"),
            bio = if (o.isNull("bio")) "" else o.optString("bio"),
            avatarUrl = avatar,
            verificationStatus = o.optString("verification_status").ifBlank { "unverified" }
        )
    }
}
